File: ios/Sources/KlinikProtocolsFeature/ProtocolsModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum ProtocolsPhase: Sendable, Equatable {
    case loading
    case loaded
    case empty
    case notPermitted
    case failed(String)
}

public struct ProtocolsState: Sendable, Equatable {
    public var phase: ProtocolsPhase = .loading
    public var documents: [ProtocolSummary] = []
    public var includeRetired = false
    public var busyId: String?
    public var error: String?

    public init() {}

    /// Documents the assistant can actually quote from today.
    public var usable: [ProtocolSummary] { documents.filter(\.isUsable) }
    /// Stored but unreachable — no embedding, or retired.
    public var unusable: [ProtocolSummary] { documents.filter { !$0.isUsable } }
}

/**
 * What the assistant is allowed to say (spec M4).
 *
 * The assistant answers only from these documents. That rule is the reason it
 * is safe to put in front of a patient at all, which makes this screen the
 * place where the clinic decides what a machine may tell somebody recovering
 * from surgery — so it shows, plainly, which documents are actually reachable.
 *
 * A document stored without an embedding is stored and invisible to retrieval.
 * Listing it beside the working ones with no distinction would let a clinic
 * believe it had answered a question it cannot answer.
 */
@MainActor
public final class ProtocolsModel {
    private let api: ProtocolsAPI
    private var state = ProtocolsState()

    public init(api: ProtocolsAPI) {
        self.api = api
    }

    public func currentState() -> ProtocolsState { state }

    public func load() async {
        do {
            state.documents = try await api.list(includeInactive: state.includeRetired)
                .sorted { $0.document.createdAt > $1.document.createdAt }
            state.phase = state.documents.isEmpty ? .empty : .loaded
        } catch APIError.forbidden {
            state.phase = .notPermitted
        } catch let error as APIError {
            state.phase = .failed(L10n.message(for: error))
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    public func showRetired(_ include: Bool) async {
        state.includeRetired = include
        state.phase = .loading
        await load()
    }

    public func add(
        title: String,
        content: String,
        procedureType: String,
        language: String
    ) async -> Bool {
        guard let problem = ProtocolsModel.problem(title: title, content: content) else {
            state.busyId = "new"
            state.error = nil

            defer { state.busyId = nil }

            do {
                _ = try await api.upload(
                    title: title.trimmed,
                    content: content.trimmed,
                    procedureType: procedureType.trimmed.isEmpty ? nil : procedureType.trimmed,
                    language: language
                )

                await load()

                return true
            } catch let error as APIError {
                state.error = L10n.message(for: error)
            } catch {
                state.error = L10n.string("error.server")
            }

            return false
        }

        state.error = problem

        return false
    }

    public func retire(_ documentId: String) async {
        state.busyId = documentId
        state.error = nil

        defer { state.busyId = nil }

        do {
            _ = try await api.retire(documentId)
            await load()
        } catch let error as APIError {
            state.error = L10n.message(for: error)
        } catch {
            state.error = L10n.string("error.server")
        }
    }

    /// Nil when the document is worth uploading.
    ///
    /// The length floor is not arbitrary: a document shorter than a sentence
    /// produces one chunk that matches everything weakly, which is how an
    /// assistant ends up quoting a heading at somebody asking about bleeding.
    public nonisolated static func problem(title: String, content: String) -> String? {
        if title.trimmed.isEmpty { return L10n.string("protocol.needTitle") }
        if content.trimmed.count < 40 { return L10n.string("protocol.needContent") }

        return nil
    }
}

private extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

File: android/feature/protocols/src/main/kotlin/xyz/klinik/feature/protocols/ProtocolsModel.kt
package xyz.klinik.feature.protocols

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.ProtocolSummary
import xyz.klinik.network.ProtocolsApi
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

sealed interface ProtocolsPhase {
    data object Loading : ProtocolsPhase
    data object Loaded : ProtocolsPhase

    /** No documents. The assistant answers nothing and sends everything on. */
    data object Empty : ProtocolsPhase

    /** Not a failure: managing the assistant's sources needs `admin.write`. */
    data object NotPermitted : ProtocolsPhase
    data class Failed(val message: UiText) : ProtocolsPhase
}

data class ProtocolsState(
    val phase: ProtocolsPhase = ProtocolsPhase.Loading,
    val documents: List<ProtocolSummary> = emptyList(),
    val showRetired: Boolean = false,
    val busy: Boolean = false,
    val error: UiText? = null,
) {
    private val visible: List<ProtocolSummary>
        get() = if (showRetired) documents else documents.filter { it.document.isActive }

    /** What the assistant can actually retrieve. */
    val usable: List<ProtocolSummary> get() = visible.filter { it.isUsable }

    /**
     * Accepted and unreachable.
     *
     * Its own list rather than a badge in the first one: a document the
     * assistant cannot see is, for the clinic's purposes, a document that was
     * never uploaded — and it looks identical to one that works.
     */
    val unusable: List<ProtocolSummary> get() = visible.filterNot { it.isUsable }
}

/** Why an upload was refused before it was sent. */
sealed interface ProtocolDraftProblem {
    data object NoTitle : ProtocolDraftProblem
    data object TooShort : ProtocolDraftProblem

    val stringKey: String
        get() = when (this) {
            NoTitle -> "protocol.needTitle"
            TooShort -> "protocol.needContent"
        }
}

/**
 * What the assistant is allowed to answer from (spec M4).
 *
 * The assistant answers only from these documents and hands everything else to
 * a person, so this list is the whole of what it knows. Two things follow.
 * A clinic with no documents is told that its assistant will forward every
 * question rather than being shown an empty list. And a document the server
 * stored but could not index is kept apart, because it is indistinguishable
 * from a working one and the assistant has never read it.
 */
class ProtocolsModel(private val api: ProtocolsApi) {
    private val _state = MutableStateFlow(ProtocolsState())
    val state: StateFlow<ProtocolsState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = ProtocolsPhase.Loading)

        try {
            val documents = api.all().sortedByDescending { it.document.createdAt }

            _state.value = _state.value.copy(
                phase = if (documents.isEmpty()) ProtocolsPhase.Empty else ProtocolsPhase.Loaded,
                documents = documents,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                phase = if (error is ApiError.Forbidden) {
                    ProtocolsPhase.NotPermitted
                } else {
                    ProtocolsPhase.Failed(messageFor(error))
                },
            )
        }
    }

    fun showRetired(show: Boolean) {
        _state.value = _state.value.copy(showRetired = show)
    }

    /**
     * Refuses a document too short to be worth retrieving.
     *
     * A two-word protocol still embeds, and the assistant will cite it — so
     * the check is here, before the upload, rather than as a surprise in an
     * answer to a patient.
     */
    fun check(title: String, content: String): ProtocolDraftProblem? = when {
        title.isBlank() -> ProtocolDraftProblem.NoTitle
        content.trim().length < 80 -> ProtocolDraftProblem.TooShort
        else -> null
    }

    suspend fun upload(
        title: String,
        content: String,
        procedureType: String?,
        language: String = "tr",
    ): Boolean {
        check(title, content)?.let { problem ->
            _state.value = _state.value.copy(error = UiText.Key(problem.stringKey))
            return false
        }

        _state.value = _state.value.copy(busy = true, error = null)

        return try {
            val added = api.upload(title.trim(), content.trim(), procedureType?.ifBlank { null }, language)

            _state.value = _state.value.copy(
                phase = ProtocolsPhase.Loaded,
                documents = listOf(added) + _state.value.documents,
                busy = false,
            )

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busy = false, error = messageFor(error))
            false
        }
    }

    /** Withdraws one, so the assistant stops answering from it. */
    suspend fun retire(documentId: String): Boolean {
        _state.value = _state.value.copy(busy = true, error = null)

        return try {
            api.remove(documentId)

            val remaining = _state.value.documents.filterNot { it.document.id == documentId }

            _state.value = _state.value.copy(
                phase = if (remaining.isEmpty()) ProtocolsPhase.Empty else ProtocolsPhase.Loaded,
                documents = remaining,
                busy = false,
            )

            true
        } catch (error: Throwable) {
            _state.value = _state.value.copy(busy = false, error = messageFor(error))
            false
        }
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}
